File: src/research_artifact_uploader/collect.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
import glob

from .manifest import ArtifactConfig, Manifest
from .paths import validate_artifact_name, should_skip_default_excluded


@dataclass
class CollectedArtifact:
    name: str
    type: str
    source_path: str
    required: bool
    status: str
    matched_paths: list[str] = field(default_factory=list)

    @property
    def matched_count(self) -> int:
        return len(self.matched_paths)


def _normalize_match_list(paths: list[Path], root: Path, artifact: ArtifactConfig, allow_outside_project: bool) -> list[Path]:
    normalized: list[Path] = []
    for p in paths:
        pp = p if p.is_absolute() else (root / p).resolve()
        if not allow_outside_project and not pp.is_relative_to(root):
            raise ValueError(f"artifact path '{artifact.path}' resolves outside cwd; use --allow-outside-project")
        if should_skip_default_excluded(artifact.path, pp, root):
            continue
        normalized.append(pp)
    return normalized


def _glob_matches(pattern: str, root: Path) -> list[Path]:
    matches = sorted(glob.glob(str(pattern)))
    return [Path(x) for x in matches]


def collect_artifacts(
    manifest: Manifest,
    project_root: Path,
    allow_outside_project: bool = False,
) -> tuple[list[CollectedArtifact], list[str]]:
    collected: list[CollectedArtifact] = []
    warnings: list[str] = []

    for artifact in manifest.artifacts:
        validate_artifact_name(artifact.name)
        raw = Path(artifact.path)
        resolved = raw if raw.is_absolute() else (project_root / raw).resolve()

        if not allow_outside_project:
            if artifact.type != "glob" and not resolved.is_relative_to(project_root):
                raise ValueError(
                    f"artifact path '{artifact.path}' is outside current working directory"
                )

        if artifact.type == "file":
            if not resolved.exists():
                if artifact.required:
                    raise FileNotFoundError(f"required file artifact missing: {artifact.path}")
                warnings.append(f"optional file artifact missing: {artifact.path}")
                status = "missing"
                matched: list[Path] = []
            elif not resolved.is_file():
                raise ValueError(f"file artifact must point to a regular file: {artifact.path}")
            else:
                status = "included"
                matched = [resolved]

        elif artifact.type == "directory":
            if not resolved.exists():
                if artifact.required:
                    raise FileNotFoundError(f"required directory artifact missing: {artifact.path}")
                warnings.append(f"optional directory artifact missing: {artifact.path}")
                status = "missing"
                matched = []
            elif not resolved.is_dir():
                raise ValueError(f"directory artifact must point to a directory: {artifact.path}")
            else:
                status = "included"
                matched = [resolved]

        elif artifact.type == "glob":
            paths = _glob_matches(artifact.path, project_root)
            if not paths:
                if artifact.required:
                    raise FileNotFoundError(f"required glob artifact has no matches: {artifact.path}")
                warnings.append(f"optional glob artifact has no matches: {artifact.path}")
                status = "missing"
                matched = []
            else:
                matched = _normalize_match_list(paths, project_root, artifact, allow_outside_project)
                status = "included"

        else:
            raise ValueError(f"unsupported artifact type: {artifact.type}")

        collected.append(
            CollectedArtifact(
                name=artifact.name,
                type=artifact.type,
                source_path=artifact.path,
                required=artifact.required,
                status=status,
                matched_paths=[str(p) for p in matched],
            )
        )

    return collected, warnings

File: src/collect.rs
use std::{collections::HashSet, fs, path::{Path, PathBuf}};

use glob::glob;

use crate::{AppResult, manifest::Manifest, paths::{validate_artifact_name, should_skip_default_excluded}};

#[derive(Debug, Clone)]
pub struct CollectedArtifact {
    pub name: String,
    pub artifact_type: String,
    pub source_path: String,
    pub required: bool,
    pub status: ArtifactStatus,
    pub matched_paths: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ArtifactStatus {
    Included,
    Missing,
}

impl CollectedArtifact {
    pub fn matched_count(&self) -> usize {
        self.matched_paths.len()
    }
}

pub fn collect_artifacts(
    manifest: &Manifest,
    project_root: &Path,
    allow_outside_project: bool,
) -> AppResult<(Vec<CollectedArtifact>, Vec<String>)> {
    let mut collected = Vec::new();
    let mut warnings = Vec::new();

    for artifact in manifest.artifacts.iter() {
        validate_artifact_name(&artifact.name)?;

        let path_text = artifact.path.as_str();
        let source = if Path::new(path_text).is_absolute() {
            PathBuf::from(path_text)
        } else {
            project_root.join(path_text)
        };

        if artifact.artifact_type != "glob" {
            if !allow_outside_project {
                if !is_within_root(&source, project_root) {
                    return Err(format!(
                        "artifact path '{}' is outside current working directory",
                        artifact.path
                    ));
                }
            }
            if should_skip_default_excluded(path_text, &source, project_root) {
                warnings.push(format!(
                    "artifact '{}' is in default-excluded path and will not be packaged",
                    artifact.path
                ));
                continue;
            }
        }

        let (status, matches) = match artifact.artifact_type.as_str() {
            "file" => collect_file(path_text, &source, artifact.required, &mut warnings)?,
            "directory" => collect_directory(path_text, &source, artifact.required, &mut warnings)?,
            "glob" => collect_glob(
                path_text,
                project_root,
                &source,
                artifact.required,
                allow_outside_project,
                &mut warnings,
            )?,
            _ => {
                return Err(format!(
                    "unsupported artifact type '{}'",
                    artifact.artifact_type
                ))
            }
        };

        collected.push(CollectedArtifact {
            name: artifact.name.clone(),
            artifact_type: artifact.artifact_type.clone(),
            source_path: artifact.path.clone(),
            required: artifact.required,
            status,
            matched_paths: matches,
        });
    }

    Ok((collected, warnings))
}

fn collect_file(path_text: &str, source: &Path, required: bool, warnings: &mut Vec<String>) -> AppResult<(ArtifactStatus, Vec<String>)> {
    if !source.exists() {
        if required {
            return Err(format!("required file artifact missing: {path_text}"));
        }
        warnings.push(format!("optional file artifact missing: {path_text}"));
        return Ok((ArtifactStatus::Missing, Vec::new()));
    }
    if !source.is_file() {
        return Err(format!("file artifact must point to a regular file: {path_text}"));
    }
    Ok((ArtifactStatus::Included, vec![source.to_string_lossy().to_string()]))
}

fn collect_directory(path_text: &str, source: &Path, required: bool, warnings: &mut Vec<String>) -> AppResult<(ArtifactStatus, Vec<String>)> {
    if !source.exists() {
        if required {
            return Err(format!("required directory artifact missing: {path_text}"));
        }
        warnings.push(format!("optional directory artifact missing: {path_text}"));
        return Ok((ArtifactStatus::Missing, Vec::new()));
    }
    if !source.is_dir() {
        return Err(format!("directory artifact must point to a directory: {path_text}"));
    }
    Ok((ArtifactStatus::Included, vec![source.to_string_lossy().to_string()]))
}

fn collect_glob(
    path_text: &str,
    project_root: &Path,
    source: &Path,
    required: bool,
    allow_outside_project: bool,
    warnings: &mut Vec<String>,
) -> AppResult<(ArtifactStatus, Vec<String>)> {
    let pattern = if source.is_absolute() {
        source.to_string_lossy().to_string()
    } else {
        project_root.join(source).to_string_lossy().to_string()
    };

    let mut candidates: Vec<PathBuf> = glob(&pattern)
        .map_err(|err| format!("invalid glob pattern '{}': {err}", path_text))?
        .filter_map(Result::ok)
        .collect();

    candidates.sort();
    if candidates.is_empty() {
        if required {
            return Err(format!("required glob artifact has no matches: {path_text}"));
        }
        warnings.push(format!("optional glob artifact has no matches: {path_text}"));
        return Ok((ArtifactStatus::Missing, Vec::new()));
    }

    let mut normalized = HashSet::<PathBuf>::new();
    let mut out = Vec::new();
    for candidate in candidates {
        if !candidate.exists() {
            continue;
        }
        let abs = if candidate.is_absolute() {
            candidate
        } else {
            project_root.join(&candidate)
        };

        if !allow_outside_project && !is_within_root(&abs, project_root) {
            return Err(format!(
                "artifact path '{}' is outside current working directory",
                path_text
            ));
        }
        if should_skip_default_excluded(path_text, &abs, project_root) {
            continue;
        }

        if normalized.insert(abs.clone()) {
            out.push(abs.to_string_lossy().to_string());
        }
    }

    if out.is_empty() {
        if required {
            return Err(format!("required glob artifact has no matches: {path_text}"));
        }
        warnings.push(format!("optional glob artifact has no matches after filtering: {path_text}"));
        return Ok((ArtifactStatus::Missing, Vec::new()));
    }

    let _ = fs::metadata(source).map_err(|err| format!("cannot access glob source '{}': {err}", path_text))?;
    Ok((ArtifactStatus::Included, out))
}

fn is_within_root(path: &Path, root: &Path) -> bool {
    path.starts_with(root)
}
